#Pre-Processing :)

Memproses hasil csv agar bisa diolah menjadi model

Tugas:

1. lakukan eksplorasi data, lihat apakah ada data yang **missing/null/nan atau ambigu, dll**
2. deteksi bahasa di **judul dan abstract**, apakah ada data yang full bahasa asing, jika ada **pisahkan** data tersebut *(preprocessing dilanjut hanya untuk data b.indo atau b.indo yang campur sedikit dengan bahasa asing)*
3. beberapa abstract mengandung keywords, oleh karena itu **pisahkan keywords** tersebut dan pindahkan ke kolom keywords
4. lakukan **cleaning** text pada **judul dan abstract**, text hanya *alphabetnumerical*, hapus whitespace tetapi pertahankan angka dengan whitespace seperti 0,0 0.0, -0, - 0, 0%
5. buat kolom baru dengan nama **fulltext** yang berisi **judul, abstract, keywords** yang sudah di preprocessing
6. lakukan **stopword** untuk fulltext, buat kolom baru dengan nama **fulltext_clean**
7. lakukan **stemming** untuk _fulltext_clean, buat kolom baru dengan nama **_fulltext_clean_stemming**
8. lakukan **deteksi pos-tag** untuk _fulltext_clean, buat kolom baru dengan nama **_fulltext_clean_postag**
9. lakukan **augmentasi data synonim replacement** untuk fulltext_clean, buat kolom baru dengan nama **_fulltext_clean_synonim**
10. lakukan **augmentasi data random deletion** untuk fulltext_clean, buat kolom baru dengan nama **_fulltext_clean_deletion**

Keterangan:

* untuk nomer 1 sampai 6 dikerjakan secara individu
* untuk nomer 7 sampai 10 dilakukan berkelompok (dua orang untuk satu nomer, dibagi saja)
* lakukan tokenization atau preprocessing lainnya jika diperlukan
* beberapa preprocessing akan memakan waktu lama, jadi tidak apa-apa jika tidak semua data di preprocessing, misal [0-500] data

## Import data dari drive

In [ ]:
#install library langdetect
!pip install langdetect

In [ ]:
!pip install PySastrawi

In [ ]:
import pandas as pd
import numpy as np
import re

from numpy import NaN

# Python program to demonstrate langdetect
from langdetect import detect
from langdetect import detect_langs

from Sastrawi.StopWordRemover.StopWordRemoverFactory import StopWordRemoverFactory
from Sastrawi.Stemmer.StemmerFactory import StemmerFactory

In [ ]:
df_real = pd.read_csv('/content/drive/MyDrive/Dokumen/Tugas-Tugas/Kuliah/RISET MBKM-STKI/Scrapping Data/VersiBaru/2117051001_AbdilahAsyidiqi_extractdatapublikasi.csv')
# df = df_real.head(500)
df = df_real
df = df.drop(columns=['Unnamed: 0'])

## Exploratory Data Analysis (EDA)
apakah ada data yang missing/null/nan atau ambigu, dll??

In [ ]:
# Cek untuk memastikan apakah csv termuat kedalam dataframe
df.head(4)

divisions                                              title  \
0     ['FT1']  Determination of Volcanic Earthquake Hypocente...   
1  ['FKIP17']  PENGARUH SKILL REPRESENTASI MATEMATIS TERHADAP...   
2    ['FP20']  TOKSISITAS METIL METSULFURON HUBUNGANNYA DENGA...   
3     ['FH2']  Principles of Islamic Sharia Economic Law in F...   

                                            abstract      subjects  \
0  Indonesia became an area passed by the Pacific...        ['TN']   
1  The Effect of Mathematical Representation Skil...        ['L1']   
2  Metil metsulfuron merupakan senyawa aktif yang...  ['Q1', 'SH']   
3                             Hasil Check Similarity        ['K1']   

                                         publication  \
0                                                NaN   
1                         Jurnal Pembelajaran Fisika   
2                                                NaN   
3  American Journal of Humanities and Social Scie...   

                                           publisher  \
0                                                NaN   
1                                Universitas Lampung   
2  Jurusan Perikanan dan Kelautan, Fakultas Perta...   
3                                             AJHSSR   

                                            creators                 date  \
0  [{'id': 'suharno@eng.unila.ac.id', 'name': {'h...                 3018   
1  [{'name': {'given': 'Illa', 'lineage': None, '...                 2107   
2  [{'name': {'honourific': None, 'lineage': None...  2024-01-15 00:00:00   
3  [{'name': {'honourific': None, 'family': 'APRI...                 2023   

  keywords                                                uri  
0      NaN  http://repository.lppm.unila.ac.id/id/eprint/5...  
1      NaN  http://repository.lppm.unila.ac.id/id/eprint/5...  
2      NaN  http://repository.lppm.unila.ac.id/id/eprint/4...  
3      NaN  http://repository.lppm.unila.ac.id/id/eprint/5...

In [ ]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 37764 entries, 0 to 37763
Data columns (total 10 columns):
 #   Column       Non-Null Count  Dtype 
---  ------       --------------  ----- 
 0   divisions    33244 non-null  object
 1   title        37764 non-null  object
 2   abstract     29593 non-null  object
 3   subjects     37746 non-null  object
 4   publication  19004 non-null  object
 5   publisher    29695 non-null  object
 6   creators     37741 non-null  object
 7   date         29296 non-null  object
 8   keywords     962 non-null    object
 9   uri          37764 non-null  object
dtypes: object(10)
memory usage: 2.9+ MB


In [ ]:
#Cek ada berapa data yang NA
df.isna().sum()

divisions       4520
title              0
abstract        8171
subjects          18
publication    18760
publisher       8069
creators          23
date            8468
keywords       36802
uri                0
dtype: int64

In [ ]:
#Cek ada berapa data yang memiliki value
df.notna().sum()

divisions      33244
title          37764
abstract       29593
subjects       37746
publication    19004
publisher      29695
creators       37741
date           29296
keywords         962
uri            37764
dtype: int64

1. Setelah di cek, ternyata masih banyak data **Null/Kosong**
2. Kolom **title** dan **subjects** tidak memiliki value **NA**, sedangkan kolom lain terdapat value **NA**
3. Cek masing-masing value yang tidak null/NA dari tiap kolom

In [ ]:
divisions_aja = df['divisions']
pd.DataFrame(divisions_aja.dropna())

divisions
0         ['FT1']
1      ['FKIP17']
2        ['FP20']
3         ['FH2']
4      ['FKIP24']
...           ...
37759     ['FT1']
37760     ['FT1']
37761   ['FKIP9']
37762   ['FKIP9']
37763     ['FH3']

[33244 rows x 1 columns]

In [ ]:
title_aja = df['title']
pd.DataFrame(title_aja.dropna())

title
0      Determination of Volcanic Earthquake Hypocente...
1      PENGARUH SKILL REPRESENTASI MATEMATIS TERHADAP...
2      TOKSISITAS METIL METSULFURON HUBUNGANNYA DENGA...
3      Principles of Islamic Sharia Economic Law in F...
4      Media Pembelajaran Digital sebagai Sumber Bela...
...                                                  ...
37759  Analysis of gravity anomaly for groundwater ba...
37760  Preliminary study of HVSR forward modeling: pa...
37761  The Effectiveness of Students’ Worksheet of   ...
37762  The Effectiveness of Students’ Worksheet of   ...
37763  Law Enforcement and Justice:  Perspective of A...

[37764 rows x 1 columns]

In [ ]:
abstract_aja = df['abstract']
pd.DataFrame(abstract_aja.dropna())

abstract
0      Indonesia became an area passed by the Pacific...
1      The Effect of Mathematical Representation Skil...
2      Metil metsulfuron merupakan senyawa aktif yang...
3                                 Hasil Check Similarity
4      Kegiatan pembelajaran siswa SMA dan SMK mutlak...
...                                                  ...
37752  Dalam perkembangannya, karya cipta hasil kreas...
37753  Indonesia is known for its cultural diversity ...
37756  Suoh merupakan daerah tektonik aktif yang meng...
37759  The research aims to delineate the groundwater...
37760  The application of horizontal to a vertical sp...

[29593 rows x 1 columns]

4. Pada bagian abstract terdapat Kata kunci yang seharusnya masuk kebagian keyword

In [ ]:
subjects_aja = df['subjects']
pd.DataFrame(subjects_aja.dropna())

subjects
0            ['TN']
1            ['L1']
2      ['Q1', 'SH']
3            ['K1']
4            ['LA']
...             ...
37759  ['QC', 'QE']
37760  ['QE', 'T1']
37761        ['L1']
37762        ['L1']
37763        ['K1']

[37746 rows x 1 columns]

In [ ]:
publication_aja = df['publication']
pd.DataFrame(publication_aja.dropna())

publication
1                             Jurnal Pembelajaran Fisika
3      American Journal of Humanities and Social Scie...
4                      EDUKATIF : JURNAL ILMU PENDIDIKAN
6                                                Heliyon
8      4th International Conference in Agroforestry 2022
...                                                  ...
37754         Research on Humanities and Social Sciences
37756                     POSITRON Berkala Ilmiah Fisika
37759  IOP Conf. Series: Journal of Physics: Conf. Se...
37760              Journal of Physics: Conference Series
37763                      Jurnal Magister Hukum Udayana

[19004 rows x 1 columns]

In [ ]:
publisher_aja = df['publisher']
pd.DataFrame(publisher_aja.dropna())

publisher
1                                    Universitas Lampung
2      Jurusan Perikanan dan Kelautan, Fakultas Perta...
3                                                 AJHSSR
4                   Universitas Pahlawan Tuanku Tambusai
5      Jurusan Pendidikan MIPA, FKIP Universitas Lampung
...                                                  ...
37757  Jurusan Teknik Geofisika, Fakultas Teknik, Uni...
37758  Jurusan Fisika, Fakultas Matematika dan Ilmu P...
37760                                     IOP Publishing
37761  International Journal of Advanced Engineering,...
37762  International Journal of Advanced Engineering,...

[29695 rows x 1 columns]

In [ ]:
date_aja = df['date']
pd.DataFrame(date_aja.dropna())

date
0                     3018
1                     2107
2      2024-01-15 00:00:00
3                     2023
4               2023-02-28
...                    ...
29291                   20
29292           0017-04-17
29293           0016-03-22
29294              0016-05
29295           0015-10-15

[29296 rows x 1 columns]

5. Value pada tahun terdapat beberapa anomali seperti 2107, 3018, 0020-12, dll

In [ ]:
keyword_aja = df['keywords']
jumlah_sebelum_ekstrak = keyword_aja.notna().sum()
pd.DataFrame(keyword_aja.dropna())

keywords
32     State Space Model, Strategic Decision-Making, ...
51     batubara, Jalan, kecepatan, pengangkutan, prod...
62     Persimpangan, Geometrik, CBD, Traffic Light, A...
63     Kabupaten  Pringsewu,  Tingkat  Ancaman,  Ting...
90     Potential Spread of Covid-19, Error Ellipsoid ...
...                                                  ...
36004           cassava, starch content and strain gauge
37032  proyek, penjadwalan, keterlambatan, crash prog...
37223  Agribisnis ubi kayu, kemitraan, kesediaan peta...
37245  alat pengolahan tanah, mekanisasi pertanian, m...
37537  Biopesticide, Immune defenses, LC-MS, Mirabili...

[962 rows x 1 columns]

6. Format penulisan keyword **tidak menggunakan pola yang sama**, contohnya:

* Index ke-655 | *: Pantai Labuhan Jukung, gelombang, abrasi, fenomena, wisata*
* Index ke-2841	| *Kata kunci: Alih fungsi lahan, Hidroponik, Pertanian pekarangan, Rumah hijau*
* Index ke-4693	| *Keywords: city infrastructure, in-depth interview method, purposive sampling technique, sustainability, TBL criteria, TBL indicators.*
* Index ke-5562	| *Copper, Residue, Retainment, Tropical Soil, Waste, Zinc*
* Index ke-14008 | *x*
* Index ke-33962 | *xxxx*
* Index ke-36235 | *XXX*

## Bila terdapat kata kunci dibagian abstract, pindahkan ke kolom Keyword
7. Buat fungsi untuk mengekstrak kata kunci dari abstrak

In [ ]:
# Fungsi untuk mengekstrak kata kunci dari abstrak menggunakan regex
def extract_keywords(abstract):
    # Pola regex untuk mencari kata kunci
    keyword_pattern = r'(Keywords|Kata Kunci|key word).*(:|—)(.*$)'
    flags = re.IGNORECASE | re.MULTILINE
    try:
      match = re.search(keyword_pattern, abstract, flags)  # Mengabaikan huruf besar/kecil
    except:
      return np.nan
    if match:
      result = match.group(3).strip()
      if result == ("" or "—" or "-" or np.nan):
        return np.nan
      else:
        return result
    else:
      return np.nan

In [ ]:
# Fungsi untuk menghapus kata kunci dari abstrak menggunakan regex
def delete_keywords(abstract):
    # Pola regex untuk mencari kata kunci
    keyword_pattern = r'(?i)(Keywords|Kata Kunci|key word).*'
    try:
      flags = re.IGNORECASE | re.MULTILINE
      match = re.search(keyword_pattern, abstract, flags)  # Mengabaikan huruf besar/kecil
      if match:
         result = re.sub(keyword_pattern, '', abstract)
      else:
        result = abstract
      if result == ("" or "—" or "-" or np.nan):
        return np.nan
      else:
        return result
    except:
      result = abstract
      return result

8. Apply function ke kolom ***keywords***
9. Apply function ke kolom baru ***new abstract***

In [ ]:
# Terapkan fungsi delete_keywords & extract_keywords pada setiap baris DataFrame
new_abstract = []

for i in range(len(df['abstract'])):
  new_abstract.append(delete_keywords(df['abstract'][i]))
  if pd.isnull(df['keywords'][i]):
    df['keywords'][i] = extract_keywords(df['abstract'][i])

df['new_abstract'] = new_abstract

In [ ]:
# Tampilkan hasilnya
df

divisions                                              title  \
0         ['FT1']  Determination of Volcanic Earthquake Hypocente...   
1      ['FKIP17']  PENGARUH SKILL REPRESENTASI MATEMATIS TERHADAP...   
2        ['FP20']  TOKSISITAS METIL METSULFURON HUBUNGANNYA DENGA...   
3         ['FH2']  Principles of Islamic Sharia Economic Law in F...   
4      ['FKIP24']  Media Pembelajaran Digital sebagai Sumber Bela...   
...           ...                                                ...   
37759     ['FT1']  Analysis of gravity anomaly for groundwater ba...   
37760     ['FT1']  Preliminary study of HVSR forward modeling: pa...   
37761   ['FKIP9']  The Effectiveness of Students’ Worksheet of   ...   
37762   ['FKIP9']  The Effectiveness of Students’ Worksheet of   ...   
37763     ['FH3']  Law Enforcement and Justice:  Perspective of A...   

                                                abstract      subjects  \
0      Indonesia became an area passed by the Pacific...        ['TN']   
1      The Effect of Mathematical Representation Skil...        ['L1']   
2      Metil metsulfuron merupakan senyawa aktif yang...  ['Q1', 'SH']   
3                                 Hasil Check Similarity        ['K1']   
4      Kegiatan pembelajaran siswa SMA dan SMK mutlak...        ['LA']   
...                                                  ...           ...   
37759  The research aims to delineate the groundwater...  ['QC', 'QE']   
37760  The application of horizontal to a vertical sp...  ['QE', 'T1']   
37761                                                NaN        ['L1']   
37762                                                NaN        ['L1']   
37763                                                NaN        ['K1']   

                                             publication  \
0                                                    NaN   
1                             Jurnal Pembelajaran Fisika   
2                                                    NaN   
3      American Journal of Humanities and Social Scie...   
4                      EDUKATIF : JURNAL ILMU PENDIDIKAN   
...                                                  ...   
37759  IOP Conf. Series: Journal of Physics: Conf. Se...   
37760              Journal of Physics: Conference Series   
37761                                                NaN   
37762                                                NaN   
37763                      Jurnal Magister Hukum Udayana   

                                               publisher  \
0                                                    NaN   
1                                    Universitas Lampung   
2      Jurusan Perikanan dan Kelautan, Fakultas Perta...   
3                                                 AJHSSR   
4                   Universitas Pahlawan Tuanku Tambusai   
...                                                  ...   
37759                                                NaN   
37760                                     IOP Publishing   
37761  International Journal of Advanced Engineering,...   
37762  International Journal of Advanced Engineering,...   
37763                                                NaN   

                                                creators                 date  \
0      [{'id': 'suharno@eng.unila.ac.id', 'name': {'h...                 3018   
1      [{'name': {'given': 'Illa', 'lineage': None, '...                 2107   
2      [{'name': {'honourific': None, 'lineage': None...  2024-01-15 00:00:00   
3      [{'name': {'honourific': None, 'family': 'APRI...                 2023   
4      [{'id': 'suparman.arif@fkip.unila.ac.id', 'nam...           2023-02-28   
...                                                  ...                  ...   
37759  [{'name': {'lineage': None, 'given': 'ahmad', ...                  NaN   
37760  [{'id': 'ahmad.zaenudin@eng.unila.ac.id', 'nam...                  NaN   
37761  [{'id': None, 'name': {'honourific': None, 'fa...                  NaN   
37

In [ ]:
print(df['new_abstract'].notna().sum())
print(df['abstract'].notna().sum())
print(df['new_abstract'].isna().sum())
print(df['abstract'].isna().sum())

29593
29593
8171
8171


In [ ]:
print(df['keywords'].isna().sum())
print(df['keywords'].notna().sum())

29816
7948


In [ ]:
#simpan dataset dengan kolom keywords baru
df.to_csv("/content/drive/MyDrive/Dokumen/Tugas-Tugas/Kuliah/RISET MBKM-STKI/Preprocessing/Dataset dengan keyword dan abstrak baru/df_with_new_keywords_and_abstract.csv", index=False)

## Buat raw Fulltext dari Title, New Abstract dan Keyword

10. Isi nilai NaN di Title, New Abstract dan Keyword menjadi string kosong
11. Buat function createRawFullText

In [ ]:
df = pd.read_csv('/content/drive/MyDrive/Dokumen/Tugas-Tugas/Kuliah/RISET MBKM-STKI/Preprocessing/Dataset dengan keyword dan abstrak baru/df_with_new_keywords_and_abstract.csv')

In [ ]:
df

divisions                                              title  \
0         ['FT1']  Determination of Volcanic Earthquake Hypocente...   
1      ['FKIP17']  PENGARUH SKILL REPRESENTASI MATEMATIS TERHADAP...   
2        ['FP20']  TOKSISITAS METIL METSULFURON HUBUNGANNYA DENGA...   
3         ['FH2']  Principles of Islamic Sharia Economic Law in F...   
4      ['FKIP24']  Media Pembelajaran Digital sebagai Sumber Bela...   
...           ...                                                ...   
37759     ['FT1']  Analysis of gravity anomaly for groundwater ba...   
37760     ['FT1']  Preliminary study of HVSR forward modeling: pa...   
37761   ['FKIP9']  The Effectiveness of Students’ Worksheet of   ...   
37762   ['FKIP9']  The Effectiveness of Students’ Worksheet of   ...   
37763     ['FH3']  Law Enforcement and Justice:  Perspective of A...   

                                                abstract      subjects  \
0      Indonesia became an area passed by the Pacific...        ['TN']   
1      The Effect of Mathematical Representation Skil...        ['L1']   
2      Metil metsulfuron merupakan senyawa aktif yang...  ['Q1', 'SH']   
3                                 Hasil Check Similarity        ['K1']   
4      Kegiatan pembelajaran siswa SMA dan SMK mutlak...        ['LA']   
...                                                  ...           ...   
37759  The research aims to delineate the groundwater...  ['QC', 'QE']   
37760  The application of horizontal to a vertical sp...  ['QE', 'T1']   
37761                                                NaN        ['L1']   
37762                                                NaN        ['L1']   
37763                                                NaN        ['K1']   

                                             publication  \
0                                                    NaN   
1                             Jurnal Pembelajaran Fisika   
2                                                    NaN   
3      American Journal of Humanities and Social Scie...   
4                      EDUKATIF : JURNAL ILMU PENDIDIKAN   
...                                                  ...   
37759  IOP Conf. Series: Journal of Physics: Conf. Se...   
37760              Journal of Physics: Conference Series   
37761                                                NaN   
37762                                                NaN   
37763                      Jurnal Magister Hukum Udayana   

                                               publisher  \
0                                                    NaN   
1                                    Universitas Lampung   
2      Jurusan Perikanan dan Kelautan, Fakultas Perta...   
3                                                 AJHSSR   
4                   Universitas Pahlawan Tuanku Tambusai   
...                                                  ...   
37759                                                NaN   
37760                                     IOP Publishing   
37761  International Journal of Advanced Engineering,...   
37762  International Journal of Advanced Engineering,...   
37763                                                NaN   

                                                creators                 date  \
0      [{'id': 'suharno@eng.unila.ac.id', 'name': {'h...                 3018   
1      [{'name': {'given': 'Illa', 'lineage': None, '...                 2107   
2      [{'name': {'honourific': None, 'lineage': None...  2024-01-15 00:00:00   
3      [{'name': {'honourific': None, 'family': 'APRI...                 2023   
4      [{'id': 'suparman.arif@fkip.unila.ac.id', 'nam...           2023-02-28   
...                                                  ...                  ...   
37759  [{'name': {'lineage': None, 'given': 'ahmad', ...                  NaN   
37760  [{'id': 'ahmad.zaenudin@eng.unila.ac.id', 'nam...                  NaN   
37761  [{'id': None, 'name': {'honourific': None, 'fa...                  NaN   
37

In [ ]:
df['abstract'] = df['abstract'].replace(np.nan, "")
df['new_abstract'] = df['new_abstract'].replace(np.nan, "")
df['title'] = df['title'].replace(np.nan, "")
df['keywords'] = df['keywords'].replace(np.nan, "")

In [ ]:
df

divisions                                              title  \
0         ['FT1']  Determination of Volcanic Earthquake Hypocente...   
1      ['FKIP17']  PENGARUH SKILL REPRESENTASI MATEMATIS TERHADAP...   
2        ['FP20']  TOKSISITAS METIL METSULFURON HUBUNGANNYA DENGA...   
3         ['FH2']  Principles of Islamic Sharia Economic Law in F...   
4      ['FKIP24']  Media Pembelajaran Digital sebagai Sumber Bela...   
...           ...                                                ...   
37759     ['FT1']  Analysis of gravity anomaly for groundwater ba...   
37760     ['FT1']  Preliminary study of HVSR forward modeling: pa...   
37761   ['FKIP9']  The Effectiveness of Students’ Worksheet of   ...   
37762   ['FKIP9']  The Effectiveness of Students’ Worksheet of   ...   
37763     ['FH3']  Law Enforcement and Justice:  Perspective of A...   

                                                abstract      subjects  \
0      Indonesia became an area passed by the Pacific...        ['TN']   
1      The Effect of Mathematical Representation Skil...        ['L1']   
2      Metil metsulfuron merupakan senyawa aktif yang...  ['Q1', 'SH']   
3                                 Hasil Check Similarity        ['K1']   
4      Kegiatan pembelajaran siswa SMA dan SMK mutlak...        ['LA']   
...                                                  ...           ...   
37759  The research aims to delineate the groundwater...  ['QC', 'QE']   
37760  The application of horizontal to a vertical sp...  ['QE', 'T1']   
37761                                                           ['L1']   
37762                                                           ['L1']   
37763                                                           ['K1']   

                                             publication  \
0                                                    NaN   
1                             Jurnal Pembelajaran Fisika   
2                                                    NaN   
3      American Journal of Humanities and Social Scie...   
4                      EDUKATIF : JURNAL ILMU PENDIDIKAN   
...                                                  ...   
37759  IOP Conf. Series: Journal of Physics: Conf. Se...   
37760              Journal of Physics: Conference Series   
37761                                                NaN   
37762                                                NaN   
37763                      Jurnal Magister Hukum Udayana   

                                               publisher  \
0                                                    NaN   
1                                    Universitas Lampung   
2      Jurusan Perikanan dan Kelautan, Fakultas Perta...   
3                                                 AJHSSR   
4                   Universitas Pahlawan Tuanku Tambusai   
...                                                  ...   
37759                                                NaN   
37760                                     IOP Publishing   
37761  International Journal of Advanced Engineering,...   
37762  International Journal of Advanced Engineering,...   
37763                                                NaN   

                                                creators                 date  \
0      [{'id': 'suharno@eng.unila.ac.id', 'name': {'h...                 3018   
1      [{'name': {'given': 'Illa', 'lineage': None, '...                 2107   
2      [{'name': {'honourific': None, 'lineage': None...  2024-01-15 00:00:00   
3      [{'name': {'honourific': None, 'family': 'APRI...                 2023   
4      [{'id': 'suparman.arif@fkip.unila.ac.id', 'nam...           2023-02-28   
...                                                  ...                  ...   
37759  [{'name': {'lineage': None, 'given': 'ahmad', ...                  NaN   
37760  [{'id': 'ahmad.zaenudin@eng.unila.ac.id', 'nam...                  NaN   
37761  [{'id': None, 'name': {'honourific': None, 'fa...                  NaN   
37

In [ ]:
def createRawFulltext(data):
  return data['title'] + ' ' + data['new_abstract'] + ' ' + data['keywords']

In [ ]:
fulltext_raw = []
fulltext_raw = createRawFulltext(df)
df["text_raw"] = fulltext_raw

In [ ]:
fulltext_raw.head(5)

0    Determination of Volcanic Earthquake Hypocente...
1    PENGARUH SKILL REPRESENTASI MATEMATIS TERHADAP...
2    TOKSISITAS METIL METSULFURON HUBUNGANNYA DENGA...
3    Principles of Islamic Sharia Economic Law in F...
4    Media Pembelajaran Digital sebagai Sumber Bela...
dtype: object

In [ ]:
#simpan dataset dengan kolom fulltext baru
df.to_csv("/content/drive/MyDrive/Dokumen/Tugas-Tugas/Kuliah/RISET MBKM-STKI/Preprocessing/Raw Fulltext/df_with_raw_fulltext.csv", index=False)

## Detect Language
Deteksi bahasa suatu jurnal dari kolom fulltext raw

In [ ]:
# deteksi akurasi bahasa

def language_detection_accuracy(column, index):
  acc_lang = detect_langs(column['text_raw'][index])
  return acc_lang

# deteksi bahasa
def language_detection(column, index):
  lang = detect(column['text_raw'][index])
  return lang

In [ ]:
df = pd.read_csv('/content/drive/MyDrive/Dokumen/Tugas-Tugas/Kuliah/RISET MBKM-STKI/Preprocessing/Raw Fulltext/df_with_raw_fulltext.csv')

In [ ]:
df.head(5)

divisions                                              title  \
0     ['FT1']  Determination of Volcanic Earthquake Hypocente...   
1  ['FKIP17']  PENGARUH SKILL REPRESENTASI MATEMATIS TERHADAP...   
2    ['FP20']  TOKSISITAS METIL METSULFURON HUBUNGANNYA DENGA...   
3     ['FH2']  Principles of Islamic Sharia Economic Law in F...   
4  ['FKIP24']  Media Pembelajaran Digital sebagai Sumber Bela...   

                                            abstract      subjects  \
0  Indonesia became an area passed by the Pacific...        ['TN']   
1  The Effect of Mathematical Representation Skil...        ['L1']   
2  Metil metsulfuron merupakan senyawa aktif yang...  ['Q1', 'SH']   
3                             Hasil Check Similarity        ['K1']   
4  Kegiatan pembelajaran siswa SMA dan SMK mutlak...        ['LA']   

                                         publication  \
0                                                NaN   
1                         Jurnal Pembelajaran Fisika   
2                                                NaN   
3  American Journal of Humanities and Social Scie...   
4                  EDUKATIF : JURNAL ILMU PENDIDIKAN   

                                           publisher  \
0                                                NaN   
1                                Universitas Lampung   
2  Jurusan Perikanan dan Kelautan, Fakultas Perta...   
3                                             AJHSSR   
4               Universitas Pahlawan Tuanku Tambusai   

                                            creators                 date  \
0  [{'id': 'suharno@eng.unila.ac.id', 'name': {'h...                 3018   
1  [{'name': {'given': 'Illa', 'lineage': None, '...                 2107   
2  [{'name': {'honourific': None, 'lineage': None...  2024-01-15 00:00:00   
3  [{'name': {'honourific': None, 'family': 'APRI...                 2023   
4  [{'id': 'suparman.arif@fkip.unila.ac.id', 'nam...           2023-02-28   

  keywords                                                uri  \
0      NaN  http://repository.lppm.unila.ac.id/id/eprint/5...   
1      NaN  http://repository.lppm.unila.ac.id/id/eprint/5...   
2      NaN  http://repository.lppm.unila.ac.id/id/eprint/4...   
3      NaN  http://repository.lppm.unila.ac.id/id/eprint/5...   
4      NaN  http://repository.lppm.unila.ac.id/id/eprint/4...   

                                        new_abstract  \
0  Indonesia became an area passed by the Pacific...   
1  The Effect of Mathematical Representation Skil...   
2  Metil metsulfuron merupakan senyawa aktif yang...   
3                             Hasil Check Similarity   
4  Kegiatan pembelajaran siswa SMA dan SMK mutlak...   

                                            text_raw  
0  Determination of Volcanic Earthquake Hypocente...  
1  PENGARUH SKILL REPRESENTASI MATEMATIS TERHADAP...  
2  TOKSISITAS METIL METSULFURON HUBUNGANNYA DENGA...  
3  Principles of Islamic Sharia Economic Law in F...  
4  Media Pembelajaran Digital sebagai Sumber Bela...

In [ ]:
acc_lang = []
lang = []

for i in range(len(df['text_raw'])):
  acc_lang.append(language_detection_accuracy(df,i))
  lang.append(language_detection(df,i))

In [ ]:
df['language'] = lang
df['language_accuracy'] = acc_lang

In [ ]:
df.head(5)

divisions                                              title  \
0     ['FT1']  Determination of Volcanic Earthquake Hypocente...   
1  ['FKIP17']  PENGARUH SKILL REPRESENTASI MATEMATIS TERHADAP...   
2    ['FP20']  TOKSISITAS METIL METSULFURON HUBUNGANNYA DENGA...   
3     ['FH2']  Principles of Islamic Sharia Economic Law in F...   
4  ['FKIP24']  Media Pembelajaran Digital sebagai Sumber Bela...   

                                            abstract      subjects  \
0  Indonesia became an area passed by the Pacific...        ['TN']   
1  The Effect of Mathematical Representation Skil...        ['L1']   
2  Metil metsulfuron merupakan senyawa aktif yang...  ['Q1', 'SH']   
3                             Hasil Check Similarity        ['K1']   
4  Kegiatan pembelajaran siswa SMA dan SMK mutlak...        ['LA']   

                                         publication  \
0                                                NaN   
1                         Jurnal Pembelajaran Fisika   
2                                                NaN   
3  American Journal of Humanities and Social Scie...   
4                  EDUKATIF : JURNAL ILMU PENDIDIKAN   

                                           publisher  \
0                                                NaN   
1                                Universitas Lampung   
2  Jurusan Perikanan dan Kelautan, Fakultas Perta...   
3                                             AJHSSR   
4               Universitas Pahlawan Tuanku Tambusai   

                                            creators                 date  \
0  [{'id': 'suharno@eng.unila.ac.id', 'name': {'h...                 3018   
1  [{'name': {'given': 'Illa', 'lineage': None, '...                 2107   
2  [{'name': {'honourific': None, 'lineage': None...  2024-01-15 00:00:00   
3  [{'name': {'honourific': None, 'family': 'APRI...                 2023   
4  [{'id': 'suparman.arif@fkip.unila.ac.id', 'nam...           2023-02-28   

  keywords                                                uri  \
0      NaN  http://repository.lppm.unila.ac.id/id/eprint/5...   
1      NaN  http://repository.lppm.unila.ac.id/id/eprint/5...   
2      NaN  http://repository.lppm.unila.ac.id/id/eprint/4...   
3      NaN  http://repository.lppm.unila.ac.id/id/eprint/5...   
4      NaN  http://repository.lppm.unila.ac.id/id/eprint/4...   

                                        new_abstract  \
0  Indonesia became an area passed by the Pacific...   
1  The Effect of Mathematical Representation Skil...   
2  Metil metsulfuron merupakan senyawa aktif yang...   
3                             Hasil Check Similarity   
4  Kegiatan pembelajaran siswa SMA dan SMK mutlak...   

                                            text_raw language  \
0  Determination of Volcanic Earthquake Hypocente...       en   
1  PENGARUH SKILL REPRESENTASI MATEMATIS TERHADAP...       en   
2  TOKSISITAS METIL METSULFURON HUBUNGANNYA DENGA...       id   
3  Principles of Islamic Sharia Economic Law in F...       en   
4  Media Pembelajaran Digital sebagai Sumber Bela...       id   

         language_accuracy  
0   [en:0.999996914004376]  
1   [en:0.999997164989931]  
2  [id:0.9999985998246005]  
3  [en:0.9999973467127118]  
4  [id:0.9999984649495912]

In [ ]:
#simpan dataset dengan kolom baru yaitu 'language' dan 'language_accuracy'
df_with_lang = df
df.to_csv("/content/drive/MyDrive/Dokumen/Tugas-Tugas/Kuliah/RISET MBKM-STKI/Preprocessing/Dataset dengan language/df_with_lang.csv", index= False)

In [ ]:
#Pisahkan data berbahasa indonesia dengan bahasa lain
df_id = df_with_lang[df['language'] == 'id']
df_other = df_with_lang[df['language'] != 'id']

In [ ]:
df_id.head(7)

divisions                                              title  \
2     ['FP20']  TOKSISITAS METIL METSULFURON HUBUNGANNYA DENGA...   
4   ['FKIP24']  Media Pembelajaran Digital sebagai Sumber Bela...   
7      ['FP1']  Strategi Keberlanjutan Jasa Air Bersih Berbasi...   
11    ['FP17']  Metode Sexing Spermatozoa XY dan Aplikasi Inse...   
22    ['FP25']  PENGARUH BAHAN PELLETING TERHADAP PERKECAMBAHA...   
23     ['FP8']  PENGENALAN BUNGA KRISAN POTONG DI DESA SUNGAI ...   
26    ['FP25']  EFEKTIVITAS PRIMING PADA BENIH JEGUNG DALAM ME...   

                                             abstract      subjects  \
2   Metil metsulfuron merupakan senyawa aktif yang...  ['Q1', 'SH']   
4   Kegiatan pembelajaran siswa SMA dan SMK mutlak...        ['LA']   
7   Persoalan penyediaan jasa air bersih merupakan...        ['S1']   
11                                                NaN        ['SF']   
22  Bawang merah merupakan komoditas pertanian pen...        ['S1']   
23  Tujuan kegiatan ini adalah mengenalkan \r\nkom...        ['S1']   
26  Kandungan Aluminium yang tinggi pada tanah Ult...        ['S1']   

                                          publication  \
2                                                 NaN   
4                   EDUKATIF : JURNAL ILMU PENDIDIKAN   
7                                                 NaN   
11                                                NaN   
22                                 Jurnal Agrotropika   
23  Jurnal Pengabdian Fakultas Pertanian Universit...   
26                                 Jurnal Agrotropika   

                                            publisher  \
2   Jurusan Perikanan dan Kelautan, Fakultas Perta...   
4                Universitas Pahlawan Tuanku Tambusai   
7                            LPPM Universitas Lampung   
11                                     Kusuma Dokumen   
22                                                NaN   
23             Fakultas Pertanian Universitas Lampung   
26                                 Jurnal Agrotropika   

                                             creators                 date  \
2   [{'name': {'honourific': None, 'lineage': None...  2024-01-15 00:00:00   
4   [{'id': 'suparman.arif@fkip.unila.ac.id', 'nam...           2023-02-28   
7   [{'name': {'given': 'Zainal', 'lineage': None,...           2023-09-19   
11  [{'name': {'honourific': None, 'family': 'Adhi...                 2023   
22  [{'id': 'agustiansyah.1972@fp.unila.ac.id', 'n...                 2023   
23  [{'name': {'honourific': None, 'given': 'Agust...              2023-03   
26  [{'name': {'family': 'Agustin', 'lineage': Non...           2023-10-01   

                               keywords  \
2                                   NaN   
4                                   NaN   
7                                   NaN   
11                                  NaN   
22  Bawang merah, Benih, Pelleting, TSS   
23                                  NaN   
26                                  NaN   

                                                  uri  \
2   http://repository.lppm.unila.ac.id/id/eprint/4...   
4   http://repository.lppm.unila.ac.id/id/eprint/4...   
7   http://repository.lppm.unila.ac.id/id/eprint/5...   
11  http://repository.lppm.unila.ac.id/id/eprint/5...   
22  http://repository.lppm.unila.ac.id/id/eprint/5...   
23  http://repository.lppm.unila.ac.id/id/eprint/5...   
26  http://repository.lppm.unila.ac.id/id/eprint/5...   

                                         new_abstract  \
2   Metil metsulfuron merupakan senyawa aktif yang...   
4   Kegiatan pembelajaran siswa SMA dan SMK mutlak...   
7   Persoalan penyediaan jasa air bersih merupakan...   
11                                                NaN   
22  Bawang merah merupakan komoditas pertanian pen...   
23  Tujuan kegiatan ini adalah mengenalkan \r\nkom...   
26  Kandungan Aluminium yang tinggi pada tanah Ult...   

                                             text_raw language  \
2   TOKSISITAS METI

In [ ]:
df_other.head(7)

divisions                                              title  \
0     ['FT1']  Determination of Volcanic Earthquake Hypocente...   
1  ['FKIP17']  PENGARUH SKILL REPRESENTASI MATEMATIS TERHADAP...   
3     ['FH2']  Principles of Islamic Sharia Economic Law in F...   
5  ['FKIP15']  Check Smimilarity: STUDI PENDAHULUAN: PENGEMBA...   
6  ['FKIP17']  Impacts of integrating engineering design proc...   
8         NaN  Assessing Willingness to Pay for Improved Wate...   
9     ['FP1']  Similarity test IOP COnference 2023 assessing ...   

                                            abstract subjects  \
0  Indonesia became an area passed by the Pacific...   ['TN']   
1  The Effect of Mathematical Representation Skil...   ['L1']   
3                             Hasil Check Similarity   ['K1']   
5                                                NaN   ['L1']   
6  Currently, science education systems around th...   ['L1']   
8  One of the important functions of the forest i...   ['S1']   
9  Similarity test artikel Assessing WIllingness ...   ['S1']   

                                         publication  \
0                                                NaN   
1                         Jurnal Pembelajaran Fisika   
3  American Journal of Humanities and Social Scie...   
5                                                NaN   
6                                            Heliyon   
8  4th International Conference in Agroforestry 2022   
9                                                NaN   

                                           publisher  \
0                                                NaN   
1                                Universitas Lampung   
3                                             AJHSSR   
5  Jurusan Pendidikan MIPA, FKIP Universitas Lampung   
6                                           ELSEVIER   
8  IOP Conf. Series: Earth and Environmental Science   
9  IOP Conf. Series: Earth and Environmental Science   

                                            creators        date keywords  \
0  [{'id': 'suharno@eng.unila.ac.id', 'name': {'h...        3018      NaN   
1  [{'name': {'given': 'Illa', 'lineage': None, '...        2107      NaN   
3  [{'name': {'honourific': None, 'family': 'APRI...        2023      NaN   
5  [{'id': 'abdurrahman.1968@fkip.unila.ac.id', '...        2023      NaN   
6  [{'name': {'honourific': None, 'family': 'Abdu...  2023-04-01      NaN   
8  [{'id': 'zainal.abidin@fp.unila.ac.id', 'name'...  2023-02-01      NaN   
9  [{'id': 'zainal.abidin@fp.unila.ac.id', 'name'...     2023-02      NaN   

                                                 uri  \
0  http://repository.lppm.unila.ac.id/id/eprint/5...   
1  http://repository.lppm.unila.ac.id/id/eprint/5...   
3  http://repository.lppm.unila.ac.id/id/eprint/5...   
5  http://repository.lppm.unila.ac.id/id/eprint/5...   
6  http://repository.lppm.unila.ac.id/id/eprint/5...   
8  http://repository.lppm.unila.ac.id/id/eprint/4...   
9  http://repository.lppm.unila.ac.id/id/eprint/5...   

                                        new_abstract  \
0  Indonesia became an area passed by the Pacific...   
1  The Effect of Mathematical Representation Skil...   
3                             Hasil Check Similarity   
5                                                NaN   
6  Currently, science education systems around th...   
8  One of the important functions of the forest i...   
9  Similarity test artikel Assessing WIllingness ...   

                                            text_raw language  \
0  Determination of Volcanic Earthquake Hypocente...       en   
1  PENGARUH SKILL REPRESENTASI MATEMATIS TERHADAP...       en   
3  Principles of Islamic Sharia Economic Law in F...       en   
5  Check Smimilarity: STUDI PENDAHULUAN: PENGEMBA...       en   
6  Impacts of integrating engineering design proc...       en   
8  Assessing Willingness to Pay for Improved Wate...       en   
9  Similarity test IOP COnference 2023 assessing ...       en   

    

In [ ]:
df_id.to_csv('/content/drive/MyDrive/Dokumen/Tugas-Tugas/Kuliah/RISET MBKM-STKI/Preprocessing/Dataset dengan language/id/df_id.csv', index= False)
df_other.to_csv('/content/drive/MyDrive/Dokumen/Tugas-Tugas/Kuliah/RISET MBKM-STKI/Preprocessing/Dataset dengan language/other/df_other.csv', index= False)

## Cleaning
bersihkan tanda baca

In [ ]:
def clean_text(text):
  result = text.lower()
  result = re.sub(r'[^\w\s\.,%\-]+', ' ', result) # Remove Symbol
  result = re.sub(r'[_]+', ' ', result) # Remove Underscore
  result = re.sub(r'([\d]+)[\s]*(%)|([\d]+)[\s]*(-)[\s]*([\d]+)|(-)[\s]*([\d]+)', r'\1\2\3\4\5\6\7', result) # Remove space between digit and symbol (1 - 2, 1 %)
  result = re.sub(r'([\D])[\.,]+', r'\1 ', result) # Remove dot and comma (exceeding)
  result = re.sub(r'[\.,]+([\D])', r'\1 ', result) # Remove dot and comma (preceeding)
  result = re.sub(r'([\D])[%]+', r'\1 ', result) # Remove Percent
  result = re.sub(r'([\D])-+([\D])|-+([\D])', r'\1\2\3 ', result) # Remove Strip
  result = re.sub(r'[\s]+', ' ', result) # Remove Multiple Whitespace
  result = result.strip()
  return result

In [ ]:
df = pd.read_csv('/content/drive/MyDrive/Dokumen/Tugas-Tugas/Kuliah/RISET MBKM-STKI/Preprocessing/Dataset dengan language/id/df_id.csv')

In [ ]:
df['fulltext'] = df['text_raw'].apply(lambda x: clean_text(x))

In [ ]:
df.head(5)

divisions                                              title  \
0    ['FP20']  TOKSISITAS METIL METSULFURON HUBUNGANNYA DENGA...   
1  ['FKIP24']  Media Pembelajaran Digital sebagai Sumber Bela...   
2     ['FP1']  Strategi Keberlanjutan Jasa Air Bersih Berbasi...   
3    ['FP17']  Metode Sexing Spermatozoa XY dan Aplikasi Inse...   
4    ['FP25']  PENGARUH BAHAN PELLETING TERHADAP PERKECAMBAHA...   

                                            abstract      subjects  \
0  Metil metsulfuron merupakan senyawa aktif yang...  ['Q1', 'SH']   
1  Kegiatan pembelajaran siswa SMA dan SMK mutlak...        ['LA']   
2  Persoalan penyediaan jasa air bersih merupakan...        ['S1']   
3                                                NaN        ['SF']   
4  Bawang merah merupakan komoditas pertanian pen...        ['S1']   

                         publication  \
0                                NaN   
1  EDUKATIF : JURNAL ILMU PENDIDIKAN   
2                                NaN   
3                                NaN   
4                 Jurnal Agrotropika   

                                           publisher  \
0  Jurusan Perikanan dan Kelautan, Fakultas Perta...   
1               Universitas Pahlawan Tuanku Tambusai   
2                           LPPM Universitas Lampung   
3                                     Kusuma Dokumen   
4                                                NaN   

                                            creators                 date  \
0  [{'name': {'honourific': None, 'lineage': None...  2024-01-15 00:00:00   
1  [{'id': 'suparman.arif@fkip.unila.ac.id', 'nam...           2023-02-28   
2  [{'name': {'given': 'Zainal', 'lineage': None,...           2023-09-19   
3  [{'name': {'honourific': None, 'family': 'Adhi...                 2023   
4  [{'id': 'agustiansyah.1972@fp.unila.ac.id', 'n...                 2023   

                              keywords  \
0                                  NaN   
1                                  NaN   
2                                  NaN   
3                                  NaN   
4  Bawang merah, Benih, Pelleting, TSS   

                                                 uri  \
0  http://repository.lppm.unila.ac.id/id/eprint/4...   
1  http://repository.lppm.unila.ac.id/id/eprint/4...   
2  http://repository.lppm.unila.ac.id/id/eprint/5...   
3  http://repository.lppm.unila.ac.id/id/eprint/5...   
4  http://repository.lppm.unila.ac.id/id/eprint/5...   

                                        new_abstract  \
0  Metil metsulfuron merupakan senyawa aktif yang...   
1  Kegiatan pembelajaran siswa SMA dan SMK mutlak...   
2  Persoalan penyediaan jasa air bersih merupakan...   
3                                                NaN   
4  Bawang merah merupakan komoditas pertanian pen...   

                                            text_raw language  \
0  TOKSISITAS METIL METSULFURON HUBUNGANNYA DENGA...       id   
1  Media Pembelajaran Digital sebagai Sumber Bela...       id   
2  Strategi Keberlanjutan Jasa Air Bersih Berbasi...       id   
3  Metode Sexing Spermatozoa XY dan Aplikasi Inse...       id   
4  PENGARUH BAHAN PELLETING TERHADAP PERKECAMBAHA...       id   

         language_accuracy                                           fulltext  
0  [id:0.9999985998246005]  toksisitas metil metsulfuron hubungannya denga...  
1  [id:0.9999984649495912]  media pembelajaran digital sebagai sumber bela...  
2  [id:0.9999969177170068]  strategi keberlanjutan jasa air bersih berbasi...  
3   [id:0.999998385833847]  metode sexing spermatozoa xy dan aplikasi inse...  
4  [id:0.9999981965987415]  pengaruh bahan pelleting terhadap perkecambaha...

In [ ]:
df.to_csv('/content/drive/MyDrive/Dokumen/Tugas-Tugas/Kuliah/RISET MBKM-STKI/Preprocessing/Fulltext Cleaned/df_fulltext_cleaned_id.csv', index=False)

## Stopword Removal

In [ ]:
def remove_stopword(text):
  factory = StopWordRemoverFactory()
  custom = [
    'mutlak',
    'era',
    'cenderung',
    'bantu',
    'paham',
    'dimana',
    'harap',
    'bahas',
    'definisi',
    'dimana',
    'dasar',
    'duga',
    'pakai',
    'pilih',
    'sesuai',
    'meliputi',
    'dituliskan',
    'dinyatakan',
    'keberlanjutan',
    'memperoleh',
    'diperoleh',
    'diharapkan',
    'mengharapkan',
    'berharap',
    'membahas',
    'dibahas',
    'didefinisikan',
    'mendefinisikan',
    'didalam',
    'kedalam',
    'memiliki',
    'dimiliki',
    'berdasarkan',
    'didasari',
    'berlandaskan',
    'diduga',
    'menduga',
    'menggambarkan',
    'digambarkan',
    'memakai',
    'dipakai',
    'memilih',
    'dipilih',
    'disesuaikan',
    'menyesuaikan',
    'kesesuaian',
    'diliputi',
    'menjabarkan',
    'dijabarkan',
    'penjabaran'
  ]
  stopwords = factory.get_stop_words()
  stopwords.extend(custom)

  result = []
  for x in  text.split():
    if x not in stopwords:
      result.append(x)

  return " ".join(result)

In [ ]:
df = pd.read_csv('/content/drive/MyDrive/Dokumen/Tugas-Tugas/Kuliah/RISET MBKM-STKI/Preprocessing/Fulltext Cleaned/df_fulltext_cleaned_id.csv')

In [ ]:
df['fulltext_stopword'] = df['fulltext'].apply(lambda x: remove_stopword(x))

In [ ]:
df.head(5)

divisions                                              title  \
0    ['FP20']  TOKSISITAS METIL METSULFURON HUBUNGANNYA DENGA...   
1  ['FKIP24']  Media Pembelajaran Digital sebagai Sumber Bela...   
2     ['FP1']  Strategi Keberlanjutan Jasa Air Bersih Berbasi...   
3    ['FP17']  Metode Sexing Spermatozoa XY dan Aplikasi Inse...   
4    ['FP25']  PENGARUH BAHAN PELLETING TERHADAP PERKECAMBAHA...   

                                            abstract      subjects  \
0  Metil metsulfuron merupakan senyawa aktif yang...  ['Q1', 'SH']   
1  Kegiatan pembelajaran siswa SMA dan SMK mutlak...        ['LA']   
2  Persoalan penyediaan jasa air bersih merupakan...        ['S1']   
3                                                NaN        ['SF']   
4  Bawang merah merupakan komoditas pertanian pen...        ['S1']   

                         publication  \
0                                NaN   
1  EDUKATIF : JURNAL ILMU PENDIDIKAN   
2                                NaN   
3                                NaN   
4                 Jurnal Agrotropika   

                                           publisher  \
0  Jurusan Perikanan dan Kelautan, Fakultas Perta...   
1               Universitas Pahlawan Tuanku Tambusai   
2                           LPPM Universitas Lampung   
3                                     Kusuma Dokumen   
4                                                NaN   

                                            creators                 date  \
0  [{'name': {'honourific': None, 'lineage': None...  2024-01-15 00:00:00   
1  [{'id': 'suparman.arif@fkip.unila.ac.id', 'nam...           2023-02-28   
2  [{'name': {'given': 'Zainal', 'lineage': None,...           2023-09-19   
3  [{'name': {'honourific': None, 'family': 'Adhi...                 2023   
4  [{'id': 'agustiansyah.1972@fp.unila.ac.id', 'n...                 2023   

                              keywords  \
0                                  NaN   
1                                  NaN   
2                                  NaN   
3                                  NaN   
4  Bawang merah, Benih, Pelleting, TSS   

                                                 uri  \
0  http://repository.lppm.unila.ac.id/id/eprint/4...   
1  http://repository.lppm.unila.ac.id/id/eprint/4...   
2  http://repository.lppm.unila.ac.id/id/eprint/5...   
3  http://repository.lppm.unila.ac.id/id/eprint/5...   
4  http://repository.lppm.unila.ac.id/id/eprint/5...   

                                        new_abstract  \
0  Metil metsulfuron merupakan senyawa aktif yang...   
1  Kegiatan pembelajaran siswa SMA dan SMK mutlak...   
2  Persoalan penyediaan jasa air bersih merupakan...   
3                                                NaN   
4  Bawang merah merupakan komoditas pertanian pen...   

                                            text_raw language  \
0  TOKSISITAS METIL METSULFURON HUBUNGANNYA DENGA...       id   
1  Media Pembelajaran Digital sebagai Sumber Bela...       id   
2  Strategi Keberlanjutan Jasa Air Bersih Berbasi...       id   
3  Metode Sexing Spermatozoa XY dan Aplikasi Inse...       id   
4  PENGARUH BAHAN PELLETING TERHADAP PERKECAMBAHA...       id   

         language_accuracy                                           fulltext  \
0  [id:0.9999985998246005]  toksisitas metil metsulfuron hubungannya denga...   
1  [id:0.9999984649495912]  media pembelajaran digital sebagai sumber bela...   
2  [id:0.9999969177170068]  strategi keberlanjutan jasa air bersih berbasi...   
3   [id:0.999998385833847]  metode sexing spermatozoa xy dan aplikasi inse...   
4  [id:0.9999981965987415]  pengaruh bahan pelleting terhadap perkecambaha...   

                                   fulltext_stopword  
0  toksisitas metil metsulfuron hubungannya masku...  
1  media pembelajaran digital sumber belajar maha...  
2  strategi jasa air bersih berbasis masyarakat k...  
3  metode sexing spermatozoa xy aplikasi insemina...  
4  pengaruh bahan pelleting perkecambahan benih b...

In [ ]:
#Simpan file
df.to_csv('/content/drive/MyDrive/Dokumen/Tugas-Tugas/Kuliah/RISET MBKM-STKI/Preprocessing/Stopword/df_fulltext_cleaned_stopword_id.csv', index = False)

## Stemming

In [ ]:
def stem(text):
  factory = StemmerFactory()
  stemmer = factory.create_stemmer()
  return stemmer.stem(text)

In [ ]:
df_steming = pd.read_csv('/content/drive/MyDrive/Dokumen/Tugas-Tugas/Kuliah/RISET MBKM-STKI/Preprocessing/Stopword/df_fulltext_cleaned_stopword_id.csv')

In [ ]:
df_steming.head(5)

divisions                                              title  \
0    ['FP20']  TOKSISITAS METIL METSULFURON HUBUNGANNYA DENGA...   
1  ['FKIP24']  Media Pembelajaran Digital sebagai Sumber Bela...   
2     ['FP1']  Strategi Keberlanjutan Jasa Air Bersih Berbasi...   
3    ['FP17']  Metode Sexing Spermatozoa XY dan Aplikasi Inse...   
4    ['FP25']  PENGARUH BAHAN PELLETING TERHADAP PERKECAMBAHA...   

                                            abstract      subjects  \
0  Metil metsulfuron merupakan senyawa aktif yang...  ['Q1', 'SH']   
1  Kegiatan pembelajaran siswa SMA dan SMK mutlak...        ['LA']   
2  Persoalan penyediaan jasa air bersih merupakan...        ['S1']   
3                                                NaN        ['SF']   
4  Bawang merah merupakan komoditas pertanian pen...        ['S1']   

                         publication  \
0                                NaN   
1  EDUKATIF : JURNAL ILMU PENDIDIKAN   
2                                NaN   
3                                NaN   
4                 Jurnal Agrotropika   

                                           publisher  \
0  Jurusan Perikanan dan Kelautan, Fakultas Perta...   
1               Universitas Pahlawan Tuanku Tambusai   
2                           LPPM Universitas Lampung   
3                                     Kusuma Dokumen   
4                                                NaN   

                                            creators                 date  \
0  [{'name': {'honourific': None, 'lineage': None...  2024-01-15 00:00:00   
1  [{'id': 'suparman.arif@fkip.unila.ac.id', 'nam...           2023-02-28   
2  [{'name': {'given': 'Zainal', 'lineage': None,...           2023-09-19   
3  [{'name': {'honourific': None, 'family': 'Adhi...                 2023   
4  [{'id': 'agustiansyah.1972@fp.unila.ac.id', 'n...                 2023   

                              keywords  \
0                                  NaN   
1                                  NaN   
2                                  NaN   
3                                  NaN   
4  Bawang merah, Benih, Pelleting, TSS   

                                                 uri  \
0  http://repository.lppm.unila.ac.id/id/eprint/4...   
1  http://repository.lppm.unila.ac.id/id/eprint/4...   
2  http://repository.lppm.unila.ac.id/id/eprint/5...   
3  http://repository.lppm.unila.ac.id/id/eprint/5...   
4  http://repository.lppm.unila.ac.id/id/eprint/5...   

                                        new_abstract  \
0  Metil metsulfuron merupakan senyawa aktif yang...   
1  Kegiatan pembelajaran siswa SMA dan SMK mutlak...   
2  Persoalan penyediaan jasa air bersih merupakan...   
3                                                NaN   
4  Bawang merah merupakan komoditas pertanian pen...   

                                            text_raw language  \
0  TOKSISITAS METIL METSULFURON HUBUNGANNYA DENGA...       id   
1  Media Pembelajaran Digital sebagai Sumber Bela...       id   
2  Strategi Keberlanjutan Jasa Air Bersih Berbasi...       id   
3  Metode Sexing Spermatozoa XY dan Aplikasi Inse...       id   
4  PENGARUH BAHAN PELLETING TERHADAP PERKECAMBAHA...       id   

         language_accuracy                                           fulltext  \
0  [id:0.9999985998246005]  toksisitas metil metsulfuron hubungannya denga...   
1  [id:0.9999984649495912]  media pembelajaran digital sebagai sumber bela...   
2  [id:0.9999969177170068]  strategi keberlanjutan jasa air bersih berbasi...   
3   [id:0.999998385833847]  metode sexing spermatozoa xy dan aplikasi inse...   
4  [id:0.9999981965987415]  pengaruh bahan pelleting terhadap perkecambaha...   

                                   fulltext_stopword  
0  toksisitas metil metsulfuron hubungannya masku...  
1  media pembelajaran digital sumber belajar maha...  
2  strategi jasa air bersih berbasis masyarakat k...  
3  metode sexing spermatozoa xy aplikasi insemina...  
4  pengaruh bahan pelleting perkecambahan benih b...

In [ ]:
df_steming.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 15867 entries, 0 to 15866
Data columns (total 16 columns):
 #   Column             Non-Null Count  Dtype 
---  ------             --------------  ----- 
 0   divisions          14133 non-null  object
 1   title              15867 non-null  object
 2   abstract           12296 non-null  object
 3   subjects           15864 non-null  object
 4   publication        6874 non-null   object
 5   publisher          12214 non-null  object
 6   creators           15851 non-null  object
 7   date               12316 non-null  object
 8   keywords           3418 non-null   object
 9   uri                15867 non-null  object
 10  new_abstract       12296 non-null  object
 11  text_raw           15867 non-null  object
 12  language           15867 non-null  object
 13  language_accuracy  15867 non-null  object
 14  fulltext           15867 non-null  object
 15  fulltext_stopword  15867 non-null  object
dtypes: object(16)
memory usage: 1.9+ MB


In [ ]:
# Deklarasi variabel
range_data_per_part = []
pembagi = 500
jumlah_data = len(df)
hasil = 0

# Penentuan Part
jumlah_part = jumlah_data//pembagi
sisa = (jumlah_data%pembagi)
if sisa != 0:
  jumlah_part = jumlah_part+1

while(True):

  if hasil <= jumlah_data-pembagi:
    range_data_per_part.append(hasil)
    hasil = hasil + pembagi
    if hasil >= jumlah_data:
      range_data_per_part.append(hasil)
      break
  else:
    range_data_per_part.append(hasil)
    hasil = hasil + sisa
    range_data_per_part.append(hasil)
    break

In [ ]:
for i in range_data_per_part:
  print(i, end=' ')

print('\n')
print('Jumlah index part : ', len(range_data_per_part))
print('Jumlah data per-part : ', len(range_data_per_part)-1)
print('Jumlah data per-part yang seharusnya : ', jumlah_part, '(Pastikan sama dengan Jumlah data per-part)')

panjang_part = len(range_data_per_part)-1

0 500 1000 1500 2000 2500 3000 3500 4000 4500 5000 5500 6000 6500 7000 7500 8000 8500 9000 9500 10000 10500 11000 11500 12000 12500 13000 13500 14000 14500 15000 15500 15867 

Jumlah index part :  33
Jumlah data per-part :  32
Jumlah data per-part yang seharusnya :  32 (Pastikan sama dengan Jumlah data per-part)


In [ ]:
# Bagi data frame per part agar proses stemming bisa dipotong

df_steming_part = []
for j in range(len(range_data_per_part)):
  if j == panjang_part:
    break
  df_steming_part.append([])
  for i in range(range_data_per_part[j],range_data_per_part[j+1]):
    df_steming_part[j].append(df.iloc[i])

In [ ]:
# cek tiap part
pd.DataFrame(df_steming_part[31])

divisions                                              title  \
15500      ['FP5']  PEER REVIEW JURNAL "STRUKTUR PENDAPATAN, TINGK...   
15501          NaN                    Porang untuk Masyarakat Lampung   
15502  ['FMIPA10']                Rancang Bangun Perpustakaan Digital   
15503   ['FKIP16']  Efikasi Diri, Lingkungan Keluarga dan Kebebasa...   
15504          NaN  Penatalaksanaan Penyakit Tuberkulosis dengan D...   
...            ...                                                ...   
15862      ['FH3']  Perlindungan Terhadap Korban Perang \r\ndalam ...   
15863      ['FH3']  HAK KEKAYAAN INTELEKTUAL BAGI PEKERJA SENI DI ...   
15864      ['FT1']  Analisis Struktur Patahan Daerah Suoh Mengguna...   
15865          NaN  Analisis Struktur Patahan Daerah Suoh Mengguna...   
15866          NaN  Similiaritu_Analisis Struktur Patahan Daerah S...   

                                                abstract    subjects  \
15500  Program Hutan Kemasyarakatan (HKm) merupakan s...      ['S1']   
15501  BUDI daya porang kali pertama\r\ndirintis oleh...      ['S1']   
15502  Berdasarkan surat edaran Dirjen Dikti No. 152/...    ['QA75']   
15503  Penelitian ini bertujuan untuk mengetahui peng...      ['L1']   
15504  Kelompok lanjut usia termasuk dalam populasi r...  ['RA1001']   
...                                                  ...         ...   
15862  Konflik bersenjata hingga kini masih menjadi p...      ['K1']   
15863  Dalam perkembangannya, karya cipta hasil kreas...      ['K1']   
15864  Suoh merupakan daerah tektonik aktif yang meng...      ['TN']   
15865                                                NaN      ['TN']   
15866                                                NaN      ['TN']   

                                             publication  \
15500                                                NaN   
15501  https://radarlampung.co.id/porang-untuk-masyar...   
15502                                     Jurnal Teknika   
15503  Efikasi Diri, Lingkungan Keluarga dan Kebebasa...   
15504     JK Unila JURNAL KEDOKTERAN UNIVERSITAS LAMPUNG   
...                                                  ...   
15862                                                NaN   
15863  Jurnal Pengabdian kepada Masyarakat Sakai Semb...   
15864                     POSITRON Berkala Ilmiah Fisika   
15865                                                NaN   
15866                                                NaN   

                                               publisher  \
15500  Penerbit Universitas Lampung with Indonesia Ne...   
15501                                                NaN   
15502         PPPM Politeknik Negeri Sriwijaya Palembang   
15503             FEB Universitas Muhammadiyah Pontianak   
15504                                           FK Unila   
...                                                  ...   
15862                                               ICRC   
15863  Lembaga Penelitian dan Pengabdian kepada Masya...   
15864                                                NaN   
15865  Jurusan Teknik Geofisika, Fakultas Teknik, Uni...   
15866  Jurusan Fisika, Fakultas Matematika dan Ilmu P...   

                                                creators  date  \
15500  [{'name': {'honourific': None, 'lineage': None...   NaN   
15501  [{'id': 'gundowino@gmail.com', 'name': {'famil...   NaN   
15502  [{'name': {'family': 'Winata', 'given': 'Andri...   NaN   
15503  [{'name': {'honourific': None, 'given': 'I Kom...   NaN   
15504  [{'name': {'honourific': None, 'lineage': None...   NaN   
...                                                  ...   ...   
15862  [{'id': None, 'name': {'honourific': None, 'gi...   NaN   
15863  [{'name': {'lineage': None, 'given': 'yunita m...   NaN   
15864  [{'name': {'given': 'ahmad', 'lineage': None, ...   NaN   
15865  [{'name': {'honourific': None, 'family': 'zaen...   NaN   
15866  [{'id': 'ahmad.zaenudin@eng.unila.ac.id', 'nam...   NaN   

                                

In [ ]:
print("Jumlah part yang dihasilkan : ",len(df_steming_part))

Jumlah part yang dihasilkan :  32


In [ ]:
#Lakukan proses steming per part. Atur berapa banyak data yang akan di stem pada bagian range
awal = 1
akhir = jumlah_part

for i in range(awal-1, akhir):
  df_steming_part_temp = pd.DataFrame(df_steming_part[i])
  df_steming_part_temp['text_stemmed'] = df_steming_part_temp['fulltext_stopword'].apply(lambda x : stem(x))

  #Simpan file
  df_steming_part_temp.to_csv(f'/content/drive/MyDrive/Dokumen/Tugas-Tugas/Kuliah/RISET MBKM-STKI/Preprocessing/Stemming/df_fulltext_cleaned_stopword_stemming_id_part{i+1}.csv', index=False)
  print(f"Part {i+1} Sukses")

Part 1 Sukses
Part 2 Sukses
Part 3 Sukses
Part 4 Sukses
Part 5 Sukses
Part 6 Sukses
Part 7 Sukses
Part 8 Sukses
Part 9 Sukses
Part 10 Sukses
Part 11 Sukses
Part 12 Sukses
Part 13 Sukses
Part 14 Sukses
Part 15 Sukses
Part 16 Sukses
Part 17 Sukses
Part 18 Sukses
Part 19 Sukses
Part 20 Sukses
Part 21 Sukses
Part 22 Sukses
Part 23 Sukses
Part 24 Sukses
Part 25 Sukses
Part 26 Sukses
Part 27 Sukses
Part 28 Sukses
Part 29 Sukses
Part 30 Sukses
Part 31 Sukses
Part 32 Sukses


In [ ]:
# Import libraries
import glob

# Get CSV files list from a folder
path = '/content/drive/MyDrive/Dokumen/Tugas-Tugas/Kuliah/RISET MBKM-STKI/Preprocessing/Stemming'
csv_files = glob.glob(path + "/*.csv")

# Read each CSV file into DataFrame
# This creates a list of dataframes
df_list = (pd.read_csv(file) for file in csv_files)

# Concatenate all DataFrames
df_hasil_stemming = pd.concat(df_list, ignore_index=True)

In [ ]:
df_hasil_stemming

divisions                                              title  \
0        ['FP20']  TOKSISITAS METIL METSULFURON HUBUNGANNYA DENGA...   
1      ['FKIP24']  Media Pembelajaran Digital sebagai Sumber Bela...   
2         ['FP1']  Strategi Keberlanjutan Jasa Air Bersih Berbasi...   
3        ['FP17']  Metode Sexing Spermatozoa XY dan Aplikasi Inse...   
4        ['FP25']  PENGARUH BAHAN PELLETING TERHADAP PERKECAMBAHA...   
...           ...                                                ...   
15862     ['FH3']  Perlindungan Terhadap Korban Perang \r\ndalam ...   
15863     ['FH3']  HAK KEKAYAAN INTELEKTUAL BAGI PEKERJA SENI DI ...   
15864     ['FT1']  Analisis Struktur Patahan Daerah Suoh Mengguna...   
15865         NaN  Analisis Struktur Patahan Daerah Suoh Mengguna...   
15866         NaN  Similiaritu_Analisis Struktur Patahan Daerah S...   

                                                abstract      subjects  \
0      Metil metsulfuron merupakan senyawa aktif yang...  ['Q1', 'SH']   
1      Kegiatan pembelajaran siswa SMA dan SMK mutlak...        ['LA']   
2      Persoalan penyediaan jasa air bersih merupakan...        ['S1']   
3                                                    NaN        ['SF']   
4      Bawang merah merupakan komoditas pertanian pen...        ['S1']   
...                                                  ...           ...   
15862  Konflik bersenjata hingga kini masih menjadi p...        ['K1']   
15863  Dalam perkembangannya, karya cipta hasil kreas...        ['K1']   
15864  Suoh merupakan daerah tektonik aktif yang meng...        ['TN']   
15865                                                NaN        ['TN']   
15866                                                NaN        ['TN']   

                                             publication  \
0                                                    NaN   
1                      EDUKATIF : JURNAL ILMU PENDIDIKAN   
2                                                    NaN   
3                                                    NaN   
4                                     Jurnal Agrotropika   
...                                                  ...   
15862                                                NaN   
15863  Jurnal Pengabdian kepada Masyarakat Sakai Semb...   
15864                     POSITRON Berkala Ilmiah Fisika   
15865                                                NaN   
15866                                                NaN   

                                               publisher  \
0      Jurusan Perikanan dan Kelautan, Fakultas Perta...   
1                   Universitas Pahlawan Tuanku Tambusai   
2                               LPPM Universitas Lampung   
3                                         Kusuma Dokumen   
4                                                    NaN   
...                                                  ...   
15862                                               ICRC   
15863  Lembaga Penelitian dan Pengabdian kepada Masya...   
15864                                                NaN   
15865  Jurusan Teknik Geofisika, Fakultas Teknik, Uni...   
15866  Jurusan Fisika, Fakultas Matematika dan Ilmu P...   

                                                creators                 date  \
0      [{'name': {'honourific': None, 'lineage': None...  2024-01-15 00:00:00   
1      [{'id': 'suparman.arif@fkip.unila.ac.id', 'nam...           2023-02-28   
2      [{'name': {'given': 'Zainal', 'lineage': None,...           2023-09-19   
3      [{'name': {'honourific': None, 'family': 'Adhi...                 2023   
4      [{'id': 'agustiansyah.1972@fp.unila.ac.id', 'n...                 2023   
...                                                  ...                  ...   
15862  [{'id': None, 'name': {'honourific': None, 'gi...                  NaN   
15863  [{'name': {'lineage': None, 'given': 'yunita m...                  NaN   
15864  [{'name': {'given': 'ahmad', 'lineage': None, ...                  NaN   
15

In [ ]:
df_hasil_stemming.to_csv('/content/drive/MyDrive/Dokumen/Tugas-Tugas/Kuliah/RISET MBKM-STKI/Preprocessing/df_final.csv', index = False)

## POS Tagging (ga kepake)

In [ ]:
!pip install python-crfsuite

     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 993.5/993.5 kB 8.9 MB/s eta 0:00:00


In [ ]:
!pip install nltk

In [ ]:
import nltk
from nltk.tag import CRFTagger
from nltk.tokenize import word_tokenize

import pycrfsuite

In [ ]:
nltk.download('punkt')

[nltk_data] Downloading package punkt to /root/nltk_data...
[nltk_data]   Unzipping tokenizers/punkt.zip.


True

In [ ]:
def pos_tag_word(word):
  token = word_tokenize(word)

  ct = CRFTagger()
  ct.set_model_file('/content/drive/MyDrive/Dokumen/Tugas-Tugas/Kuliah/RISET MBKM-STKI/Preprocessing/requirements/all_indo_man_tag_corpus_model.crf.tagger')
  result = ct.tag_sents([token])

  return result[0]

In [ ]:
df = pd.read_csv('/content/drive/MyDrive/Dokumen/Tugas-Tugas/Kuliah/RISET MBKM-STKI/Preprocessing/Stemming/df_fulltext_cleaned_stopword_stemming_id_part1.csv')

In [ ]:
df.head(5)

Unnamed: 0 divisions                                              title  \
0           0       NaN                              Jalan Sunyi Sang Guru   
1           1       NaN  Kewenangan Kejaksaan dalam Penyidikan dan Prap...   
2           2       NaN            Paradoks dan Epistimologi SAKSI MAHKOTA   
3           3   ['FK1']       Modul Pelatian TOT Bagi Dosen Home Based DLP   
4           4   ['FT8']  Peer Review - APLIKASI AUGMENTED REALITY NAVIG...   

                                            abstract subjects publication  \
0                                                NaN   ['K1']         NaN   
1                                                NaN   ['K1']         NaN   
2                                                NaN   ['K1']         NaN   
3                                                  -   ['R1']         NaN   
4  Aplikasi Augmented Reality Navigasi Koleksi Pe...   ['T1']         NaN   

                             publisher  date keywords  \
0                      Aura Publishing   NaN      NaN   
1                      Aura Publishing   NaN      NaN   
2                      Aura Publishing   NaN      NaN   
3                   BP SDM Kemenkes RI   NaN      NaN   
4  Fakultas Teknik Universitas Lampung   NaN      NaN   

                                        new_abstract  \
0                                                NaN   
1                                                NaN   
2                                                NaN   
3                                                  -   
4  Aplikasi Augmented Reality Navigasi Koleksi Pe...   

                                        fulltext_raw language  \
0                            Jalan Sunyi Sang Guru         id   
1  Kewenangan Kejaksaan dalam Penyidikan dan Prap...       id   
2          Paradoks dan Epistimologi SAKSI MAHKOTA         id   
3    Modul Pelatian TOT Bagi Dosen Home Based DLP -        id   
4  Peer Review - APLIKASI AUGMENTED REALITY NAVIG...       id   

                                language_accuracy  \
0                         [id:0.9999963846268605]   
1                         [id:0.9999962931569774]   
2                         [id:0.9999961712909714]   
3  [id:0.7142826049168352, en:0.2857163348048405]   
4                         [id:0.9999957742771008]   

                                            fulltext  \
0                              jalan sunyi sang guru   
1  kewenangan kejaksaan dalam penyidikan dan prap...   
2            paradoks dan epistimologi saksi mahkota   
3       modul pelatian tot bagi dosen home based dlp   
4  peer review aplikasi augmented reality navigas...   

                                      fulltext_clean  \
0                              jalan sunyi sang guru   
1  kewenangan kejaksaan penyidikan prapenuntutan ...   
2                paradoks epistimologi saksi mahkota   
3            modul pelatian tot dosen home based dlp   
4  peer review aplikasi augmented reality navigas...   

                                        text_stemmed  
0                              jalan sunyi sang guru  
1  wenang jaksa sidi prapenuntutan tindak pidana ...  
2                paradoks epistimologi saksi mahkota  
3            modul pelatian tot dosen home based dlp  
4  peer review aplikasi augmented reality navigas...

In [ ]:
df['fulltext_clean_postag'] = df['fulltext_clean'].apply(lambda x: pos_tag_word(x))
df.head(5)

Unnamed: 0 divisions                                              title  \
0           0       NaN                              Jalan Sunyi Sang Guru   
1           1       NaN  Kewenangan Kejaksaan dalam Penyidikan dan Prap...   
2           2       NaN            Paradoks dan Epistimologi SAKSI MAHKOTA   
3           3   ['FK1']       Modul Pelatian TOT Bagi Dosen Home Based DLP   
4           4   ['FT8']  Peer Review - APLIKASI AUGMENTED REALITY NAVIG...   

                                            abstract subjects publication  \
0                                                NaN   ['K1']         NaN   
1                                                NaN   ['K1']         NaN   
2                                                NaN   ['K1']         NaN   
3                                                  -   ['R1']         NaN   
4  Aplikasi Augmented Reality Navigasi Koleksi Pe...   ['T1']         NaN   

                             publisher  date keywords  \
0                      Aura Publishing   NaN      NaN   
1                      Aura Publishing   NaN      NaN   
2                      Aura Publishing   NaN      NaN   
3                   BP SDM Kemenkes RI   NaN      NaN   
4  Fakultas Teknik Universitas Lampung   NaN      NaN   

                                        new_abstract  \
0                                                NaN   
1                                                NaN   
2                                                NaN   
3                                                  -   
4  Aplikasi Augmented Reality Navigasi Koleksi Pe...   

                                        fulltext_raw language  \
0                            Jalan Sunyi Sang Guru         id   
1  Kewenangan Kejaksaan dalam Penyidikan dan Prap...       id   
2          Paradoks dan Epistimologi SAKSI MAHKOTA         id   
3    Modul Pelatian TOT Bagi Dosen Home Based DLP -        id   
4  Peer Review - APLIKASI AUGMENTED REALITY NAVIG...       id   

                                language_accuracy  \
0                         [id:0.9999963846268605]   
1                         [id:0.9999962931569774]   
2                         [id:0.9999961712909714]   
3  [id:0.7142826049168352, en:0.2857163348048405]   
4                         [id:0.9999957742771008]   

                                            fulltext  \
0                              jalan sunyi sang guru   
1  kewenangan kejaksaan dalam penyidikan dan prap...   
2            paradoks dan epistimologi saksi mahkota   
3       modul pelatian tot bagi dosen home based dlp   
4  peer review aplikasi augmented reality navigas...   

                                      fulltext_clean  \
0                              jalan sunyi sang guru   
1  kewenangan kejaksaan penyidikan prapenuntutan ...   
2                paradoks epistimologi saksi mahkota   
3            modul pelatian tot dosen home based dlp   
4  peer review aplikasi augmented reality navigas...   

                                        text_stemmed  \
0                              jalan sunyi sang guru   
1  wenang jaksa sidi prapenuntutan tindak pidana ...   
2                paradoks epistimologi saksi mahkota   
3            modul pelatian tot dosen home based dlp   
4  peer review aplikasi augmented reality navigas...   

                               fulltext_clean_postag  
0  [(jalan, NN), (sunyi, VB), (sang, NN), (guru, ...  
1  [(kewenangan, NN), (kejaksaan, NN), (penyidika...  
2  [(paradoks, NN), (epistimologi, NN), (saksi, N...  
3  [(modul, VB), (pelatian, NN), (tot, FW), (dose...  
4  [(peer, FW), (review, FW), (aplikasi, NN), (au...

In [ ]:
df = df.drop(['Unnamed: 0'], axis=1)

In [ ]:
#Save file
df.to_csv(f'/content/drive/MyDrive/Dokumen/Tugas-Tugas/Kuliah/RISET MBKM-STKI/Preprocessing/POS Tagging/df_fulltext_clean_stopword_stemming_posttag_id_part1.csv')

# Augmentasi (Perbanyak data) | (ga kepake)

## Data Augmentation - Synonim

In [ ]:
!pip install tqdm

In [ ]:
import requests
import json
from tqdm import tqdm

In [ ]:
f = requests.get('https://raw.githubusercontent.com/victoriasovereigne/tesaurus/master/dict.json')
data = json.loads(f.text)

In [ ]:
df = pd.read_csv("/content/drive/MyDrive/Dokumen/Tugas-Tugas/Kuliah/RISET MBKM-STKI/Preprocessing/POS Tagging/df_fulltext_clean_stopword_stemming_posttag_id.csv")

In [ ]:
def replace_text(text):
  result = []
  text_token = text.split(' ')
  for i in range(len(text_token)):
    if text_token[i] in data.keys():
      for sinonim in data[text_token[i]]['sinonim']:
        temp = text_token.copy()
        temp[i] = sinonim
        result.append(' '.join(temp))

  return result

In [ ]:
def augment_synonim(data):
  result = pd.DataFrame()

  for i in replace_text(data['fulltext_clean']):
    temp_df = pd.DataFrame(data).T
    temp_df['fulltext_clean_synonim'] = i
    result = pd.concat([result, temp_df], ignore_index=True)

    temp_df

  return result

In [ ]:
df_augment = pd.DataFrame()
for i in tqdm(range(len(df))):
  df_augment = pd.concat([df_augment, augment_synonim(df.iloc[i])], ignore_index=True)

df_augment

100%|██████████| 233/233 [01:46<00:00,  2.19it/s]


Unnamed: 0   divisions  \
0              1         NaN   
1              1         NaN   
2              1         NaN   
3              1         NaN   
4              1         NaN   
...          ...         ...   
84500        499  ['FMIPA4']   
84501        499  ['FMIPA4']   
84502        499  ['FMIPA4']   
84503        499  ['FMIPA4']   
84504        499  ['FMIPA4']   

                                                   title  \
0                                  Jalan Sunyi Sang Guru   
1                                  Jalan Sunyi Sang Guru   
2                                  Jalan Sunyi Sang Guru   
3                                  Jalan Sunyi Sang Guru   
4                                  Jalan Sunyi Sang Guru   
...                                                  ...   
84500  Pemanfaatan Rumput Laut (Euchema Sp) Manjadi P...   
84501  Pemanfaatan Rumput Laut (Euchema Sp) Manjadi P...   
84502  Pemanfaatan Rumput Laut (Euchema Sp) Manjadi P...   
84503  Pemanfaatan Rumput Laut (Euchema Sp) Manjadi P...   
84504  Pemanfaatan Rumput Laut (Euchema Sp) Manjadi P...   

                                                abstract subjects  \
0                                                    NaN   ['K1']   
1                                                    NaN   ['K1']   
2                                                    NaN   ['K1']   
3                                                    NaN   ['K1']   
4                                                    NaN   ['K1']   
...                                                  ...      ...   
84500  Masyarakat Desa Legundi Kecamatan Ketapang Lam...   ['Q1']   
84501  Masyarakat Desa Legundi Kecamatan Ketapang Lam...   ['Q1']   
84502  Masyarakat Desa Legundi Kecamatan Ketapang Lam...   ['Q1']   
84503  Masyarakat Desa Legundi Kecamatan Ketapang Lam...   ['Q1']   
84504  Masyarakat Desa Legundi Kecamatan Ketapang Lam...   ['Q1']   

                                        publication             publisher  \
0                                               NaN       Aura Publishing   
1                                               NaN       Aura Publishing   
2                                               NaN       Aura Publishing   
3                                               NaN       Aura Publishing   
4                                               NaN       Aura Publishing   
...                                             ...                   ...   
84500  JURNAL PENGABDIAN KEPADA MASYARAKAT TABIKPUN  Fakultas MIPA, UNILA   
84501  JURNAL PENGABDIAN KEPADA MASYARAKAT TABIKPUN  Fakultas MIPA, UNILA   
84502  JURNAL PENGABDIAN KEPADA MASYARAKAT TABIKPUN  Fakultas MIPA, UNILA   
84503  JURNAL PENGABDIAN KEPADA MASYARAKAT TABIKPUN  Fakultas MIPA, UNILA   
84504  JURNAL PENGABDIAN KEPADA MASYARAKAT TABIKPUN  Fakultas MIPA, UNILA   

      date                                           keywords  \
0      NaN                                                NaN   
1      NaN                                                NaN   
2      NaN                                                NaN   
3      NaN                                                NaN   
4      NaN                                                NaN   
...    ...                                                ...   
84500  NaN  Euchema sp, Masker Wajah, Pasta Gigi, Rumput Laut   
84501  NaN  Euchema sp, Masker Wajah, Pasta Gigi, Rumput Laut   
84502  NaN  Euchema sp, Masker Wajah, Pasta Gigi, Rumput Laut   
84503  NaN  Euchema sp, Masker Wajah, Pasta Gigi, Rumput Laut   
84504  NaN  Euchema sp, Masker Wajah, Pasta Gigi, Rumput Laut   

                                            new_abstract  \
0                                                    NaN   
1                                                    NaN   
2                                                    NaN   
3                                                    NaN   
4                                                    NaN   
...  

In [ ]:
df = df_augment

In [ ]:
#Simpan file
df.to_csv('/content/drive/MyDrive/Dokumen/Tugas-Tugas/Kuliah/RISET MBKM-STKI/Preprocessing/Augmentation - Synonim/df_fulltext_clean_stopword_stemming_posttag_sysnonim_id.csv')

## Augmentation - Random Deletion

In [ ]:
import random

In [ ]:
def delete_random(word):
  token = word_tokenize(word)
  random_element = random.choice(token)
  token.remove(random_element)

  return ' '.join(token)

In [ ]:
df_augment_random = df
df_augment_random['fulltext_clean_augment_deletion'] = df_augment_random['fulltext_clean'].apply(lambda x: delete_random(x))

In [ ]:
df_augment_random.head(10)

In [ ]:
#Simpan file
df_augment_random.to_csv('/content/drive/MyDrive/Dokumen/Tugas-Tugas/Kuliah/RISET MBKM-STKI/Preprocessing/Augmentation - Deletion/df_fulltext_clean_stopword_stemming_posttag_sysnonim_deletion_id.csv')

In [ ]:
#Simpan file
df_augment_random.to_csv('/content/drive/MyDrive/Dokumen/Tugas-Tugas/Kuliah/RISET MBKM-STKI/Preprocessing/df_final.csv')